performance_manager/lib/rt_trip_updates.py

```diff
@@ -0,0 +1,250 @@
+from typing import Optional, Union, List, Dict, Any
+
+import logging
+import pandas
+import numpy
+
+import sqlalchemy as sa
+from sqlalchemy.orm import sessionmaker
+
+from .s3_utils import read_parquet
+from .gtfs_utils import start_time_to_seconds, add_event_hash_column
+from .postgres_utils import TripUpdateEvents, MetadataLog, get_unprocessed_files
+
+
+def get_tu_dataframe(to_load: Union[str, List[str]]) -> pandas.DataFrame:
+    """
+    return a dataframe from a trip updates parquet file (or list of files)
+    with expected columns
+    """
+    trip_update_columns = [
+        "timestamp",
+        "stop_time_update",
+        "direction_id",
+        "route_id",
+        "start_date",
+        "start_time",
+        "vehicle_id",
+    ]
+    trip_update_filters = [
+        ("direction_id", "in", (0, 1)),
+        ("timestamp", ">", 0),
+        ("route_id", "!=", "None"),
+        ("start_date", "!=", "None"),
+        ("start_time", "!=", "None"),
+        ("vehicle_id", "!=", "None"),
+    ]
+
+    return read_parquet(
+        to_load, columns=trip_update_columns, filters=trip_update_filters
+    )
+
+
+def explode_stop_time_update(row: pandas.Series) -> Optional[List[dict]]:
+    """
+    function to be used with pandas `apply` dataframe method
+    explode nested list of dicts in stop_time_update column
+    """
+    append_dict = {
+        "timestamp": row["timestamp"],
+        "direction_id": row["direction_id"],
+        "route_id": row["route_id"],
+        "start_date": row["start_date"],
+        "start_time": row["start_time"],
+        "vehicle_id": row["vehicle_id"],
+    }
+    return_list: List[Dict[str, Any]] = []
+    for record in row["stop_time_update"]:
+        try:
+            arrival_time = int(record["arrival"]["time"])
+        except (TypeError, KeyError):
+            continue
+        if (
+            arrival_time - row["timestamp"] < 0
+            or arrival_time - row["timestamp"] > 120
+        ):
+            continue
```

Comment:
a comment on why the time has to be in this window would be helpful for future us.

Reply:
Will add a comment on this, the 120 value selection is kind of arbitrary. We're just filtering out predictions that are farther into the future and unlikely to be used as the final stop event prediction. 2 minutes seemed like a conservative value.